InterruptClient.rb

```diff
@@ -0,0 +1,238 @@
+  end
+
+  def set_name
+    puts START_MSG
+    print PROMPT
+    input = STDIN.gets.chomp
+    if input.length > 8 || /\W/.match(input)
+      puts NAME_FORMAT_INSTRUCTIONS
+      set_name
+    else
+      @name = input
+    end
+  end
+
+  def handshake
+    ackd = false
+    while not ackd
+      send_msg(msg_connect(@name))
+      incoming = IO.select([@client], nil, nil, HANDSHAKE_WAIT)
+      if ((not incoming.nil?) && ack?)
+        ackd = true
+      end
+    end
+  end
+
+  def ack?
+    data, sender = @client.recvfrom(MAX_MSG_LENGTH)
+    msg = parse_msg(data, sender)
+    return false if msg.nil?
+
+    msg['type'] == 'ack'
+  end
+
+  def instructions
+    Console.clear
+    puts INSTRUCTIONS
+  end
+
+  def receive_loop
+    loop do
+      to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
+      to_read[0].each { |ios|
+        if ios == @client
+          msg, sender = ios.recvfrom(MAX_MSG_LENGTH)
+          handle_msg(msg, sender)
+        elsif ios.tty? # ios comes from terminal
+          input = STDIN.getc
+          handle_key(input)
+        end
+      }
+    end
+  end
+
+  def handle_msg(msg, sender)
+    msg = parse_msg(msg, sender)
+    return if msg.nil?
+
+    case msg['type']
+    when 'chat'
+      time = msg['time']
+      return if time < @latest_chat
+
+      @latest_chat = time
+      chat_array = msg['body']
+      names_array = msg['names']
+      names_string, names_length = names_data(names_array)
+
+      indent = chat_array.length > @term_width ? 1 : (@term_width - chat_array.length) / 2
+
+      Console.cursor_pos(2, 3)
+      print names_string
+      Console.cursor_pos(TEXT_LINE, indent)
+      print chat_string(chat_array)
+    end
+  end
+
+  def chat_string(chat_array)
+    chat_array.inject('') { |string, item|
+        text, color_code = item
+        color = color_code.nil? ? nil : COLORS[color_code]
+
+        string + Console.color(color, text)
+      }
+  end
+
+  def names_data(names_array)
+    names_string = ''
+    names_length = 0 # names_string may have color encoding chars, so we can't just check its length
+
+    names_array.each {|item|
+      nickname, color_code, emph = item
+      names_length += nickname.length + 2
+
+      nickname = color_code.nil? ? nickname : Console.color(COLORS[color_code], nickname)
+      nickname = emph ? Console.emph(nickname) : nickname
+      names_string += " #{nickname} "
+      }
+
+    return [names_string, names_length]
+  end
+
+  # disregard eg return, delete, backspace keys presses
+  # only pass on to server word, punctuation, or space characters
+  def handle_key(input)
+    if input == CMD_QUIT
+      bye
+    elsif not /^[[[:word:]][[:punct:]] ]$/.match(input).nil?
+      send_msg(msg_chat(input))
+    end
+  end
+
+  # return parsed msg if correct format and comes from Interrupt server
+  # return nil otherwise
+  def parse_msg(data, sender)
+    return nil if not is_server?(sender)
```

Comment:
`return unless is_server?(sender)` is a bit clearer, although if you could do `return unless sender.is_server?` then it would be English!